mars/services/web/core.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 1999-2020 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cloudpickle
+import inspect
+import logging
+import requests
+import sys
+import threading
+import traceback
+
+from tornado.httpclient import AsyncHTTPClient
+from .supervisor import MarsRequestHandler
+
+logger = logging.getLogger(__name__)
+
+
+def serialize(obj):
+    from mars.serialization import serialize
+    return cloudpickle.dumps(serialize(obj))
+
+
+def deserialize(binary):
+    from mars.serialization import deserialize
+    return deserialize(*(cloudpickle.loads(binary)))
+
+
+_web_address_local = threading.local()
+
+
+def set_web_address(web_address):
+    _web_address_local.web_address = web_address
+
+
+def get_web_address():
+    assert hasattr(_web_address_local, 'web_address'), 'Please set web_address first before get web address'
+    return _web_address_local.web_address
+
+
+class _HandlerException(Exception):
+
+    def __init__(self, exc_type, exc_value: BaseException, exc_traceback):
+        self.exc_type = exc_type
+        self.exc_value = exc_value
+        self.exc_traceback = exc_traceback
+
+
+class ServiceWebHandlerBase(MarsRequestHandler):
+    _api_instances = dict()
+    _api_cls = None
+
+    async def post(self, path, **_):
+        api_method_name = path
+        # params format: api_id[None], args, kwargs
+        api_id, args, kwargs = deserialize(self.request.body)
+        try:
+            if not api_id:
+                if hasattr(self, api_method_name):
+                    method = getattr(self, api_method_name)
+                else:
+                    method = getattr(self._api_cls, api_method_name)
+            else:
+                # call method of api instance
+                method = getattr(self._api_instances[api_id], api_method_name)
+            if kwargs:
+                # Some methods are decorated with @alru_cache, which doesn't support dict as part of cache key.
+                result = method(*args, **kwargs)
+            else:
+                result = method(*args)
+            if inspect.iscoroutine(result):
+                result = await result
+            self.write(serialize(result))
+        except Exception as e:
+            logger.info(f'Execute method {api_method_name} with {api_id, args, kwargs} failed, got exception {e}')
+            traceback.format_exc()
```

Comment:
What is `traceback.format_exc()` for?

Reply:
The exception hanppens in the server. Print the traceback can make troubleshooting much more easy.

Reply:
`traceback.format_exc()` does not print the exception. Please refer to the Python doc: https://docs.python.org/3/library/traceback.html#traceback.format_exc

Reply:
My mistake, changed to `print_exc`